### PR TITLE
Implement image deletion

### LINKED ISSUE #3033: Implement image delete
Created a project image and promoted it to silo. When trying to delete it, I got a 500 error:
```
pisces-2:fio angela$ oxide image list
success
ImageResultsPage {
    items: [
        Image {
            block_size: ByteCount(
                512,
            ),
            description: "ubuntu 20.04 LTS (focal)",
            digest: None,
            id: 129c2be0-6e23-49ac-a1af-f4b8322b0a97,
            name: Name(
                "ubuntu20",
            ),
            os: "ubuntu",
            project_id: None,
            size: ByteCount(
                3221225472,
            ),
            time_created: 2023-05-06T03:23:26.804978Z,
            time_modified: 2023-05-06T03:24:21.781153Z,
            url: None,
            version: "20.04 LTS",
        },
    ],
    next_page: Some(
        "eyJ2IjoidjEiLCJwYWdlX3N0YXJ0Ijp7InNvcnRfYnkiOiJuYW1lX2FzY2VuZGluZyIsInByb2plY3QiOm51bGwsImluY2x1ZGVfc2lsb19pbWFnZXMiOm51bGwsImxhc3Rfc2VlbiI6InVidW50dTIwIn19",
    ),
}
pisces-2:fio angela$ oxide image delete --image ubuntu20
error
Error Response: status: 500 Internal Server Error; headers: {"content-type": "application/json", "x-request-id": "679b0c3e-b715-4316-bf91-75ead9f7cd28", "content-length": "124", "date": "Sat, 06 May 2023 19:26:29 GMT"}; value: Error { error_code: Some("Internal"), message: "Internal Server Error", request_id: "679b0c3e-b715-4316-bf91-75ead9f7cd28" }
```
Nexus log:
```
19:26:30.533Z INFO 7f1d61df-321c-4b5b-bc92-10c48de4c432 (dropshot_external): request completed
    error_message_external = Internal Server Error
    error_message_internal = endpoint is not implemented
    local_addr = 172.30.1.5:80
    method = DELETE
    remote_addr = 172.20.17.42:53461
    req_id = 679b0c3e-b715-4316-bf91-75ead9f7cd28
    response_code = 500
    uri = //v1/images/ubuntu20
```

### PR DESCRIPTION
**Pull request by jmpesp**
Also, make sure that multiple volume construction requests can only be copied for read-only Regions: the `region` table has a column for the volume that owns it, and this means there's a one to one mapping for the volume that references the region read-write. Ensure this in `volume_checkout_randomize_ids` otherwise multiple volumes could reference the same regions and this could lead to accounting / cleanup issues.

Fixes #3033

### CODE REVIEW

File: nexus/tests/integration_tests/projects.rs [hunk 1]
```diff
@@ -245,32 +245,14 @@ async fn test_project_deletion_with_image(cptestctx: &ControlPlaneTestContext) {
...
-    // NexusRequest::object_delete(&client, &image_url)
-    //     .authn_as(AuthnMode::PrivilegedUser)
-    //     .execute()
-    //     .await
-    //     .expect("failed to delete image");
+    NexusRequest::object_delete(&client, &image_url)
+        .authn_as(AuthnMode::PrivilegedUser)
+        .execute()
+        .await
+        .expect("failed to delete image");
```
Comment (david-crespo) [inline nexus/tests/integration_tests/projects.rs:253]: I would add a check here that an attempt to fetch the image now 404s

File: nexus/tests/integration_tests/volume_management.rs [hunk 2]
```diff
@@ -1114,7 +1112,210 @@ async fn test_create_image_from_snapshot_delete(
...
+                    .await
+                    .expect("failed to delete snapshot");
+            }
+        }
+    }
+
+    // Assert everything was cleaned up
+    assert!(disk_test.crucible_resources_deleted().await);
+}
+
```
Comment (david-crespo) [inline nexus/tests/integration_tests/volume_management.rs:1242]: I'd find it helpful to see a comment here about why you're testing with different orders.

**Review by david-crespo — APPROVED** [review]
Looks great pending my two small comments.

File: nexus/src/app/image.rs [hunk 3]
```diff
@@ -356,26 +358,33 @@ impl super::Nexus {
...
             ImageLookup::ProjectImage(lookup) => {
-                lookup.lookup_for(authz::Action::Delete).await?;
+                let (_, _, authz_image, image) =
+                    lookup.fetch_for(authz::Action::Delete).await?;
+                sagas::image_delete::ImageParam::Project { authz_image, image }
             }
             ImageLookup::SiloImage(lookup) => {
-                lookup.lookup_for(authz::Action::Delete).await?;
+                let (_, authz_image, image) =
+                    lookup.fetch_for(authz::Action::Delete).await?;
```
Comment (david-crespo) [inline nexus/src/app/image.rs:374]: These permissions checks seem to me the natural thing, but I want to make sure they actually work as you would expect, i.e., silo images would require stronger permissions to delete than project images. I believe the fact that `SiloImage` is a child of the silo but `ProjectImage` is a child of the project gets you all this for free. Unfortunately this is a little tedious to write tests around.

https://github.com/oxidecomputer/omicron/blob/463cc1ae256a5e22b9726ae58ea8b651e81e3882/nexus/db-queries/src/db/lookup.rs#L581-L588

https://github.com/oxidecomputer/omicron/blob/463cc1ae256a5e22b9726ae58ea8b651e81e3882/nexus/db-queries/src/db/lookup.rs#L659-L666

Comment (jmpesp) [reply nexus/src/app/image.rs:374]: I agree, I added a test in 2856f165a

File: nexus/tests/integration_tests/volume_management.rs (hunk 2 above)
Comment (jmpesp) [reply nexus/tests/integration_tests/volume_management.rs:1242]: +1, done in bef3de54e

File: nexus/tests/integration_tests/projects.rs (hunk 1 above)
Comment (jmpesp) [reply nexus/tests/integration_tests/projects.rs:253]: done in bef3de54e

**Review by david-crespo — APPROVED** [review]
Perfect test. Double approved!

**Pull request merged**